### PR TITLE
make credential_identifiers mandatory for authorization_details flow

### LINKED ISSUE #294: Potential improvements for the big picture of issuance flows
I've tried to summarize all issuance options regarding the use of:
- authorization_details / scope
- credential_identifiers

OpenID4VCI_v3

It seems that there are many possible combinations to do issuance with OpenID4VCI, in my opinion it may be too many and we should consider consolidation.

My observations:
- `scope` is an easy/uncomplicated way to do most use cases :+1:
  - no support for different Credential Datasets but thats ok
  - it is weird that Credential Request uses `format` +format specific parameters instead of `scope`  :thinking: 
- authorization_details enables the Wallet to choose between different Credential Datasets :+1:
  - authorization_details has two options: `credential_configuration_id`  vs `format` + format specific parameter :thinking:
  - when authorization_details is used `credential_identifiers` is still optional :thinking:
  - this means that Credential Request depends on the Token Response :-1:
- in preAuth Code Flow, we have a third option, providing `credential_configuration_ids`, but probably fine
  - its unclear which Credential the Wallet is requesting if multiple IDs are provided
  - `credential_identifiers` cannot be used, because it is not authorization_details flow :-1:
  - Credential Datasets cannot be requested :-1:

Proposals:
1: make `credential_identifiers` mandatory
2: enable `credential_identifiers` for Pre-Auth Code Flow
3: enable `authorization_details` with `credential_configuration_id` in Token Request for Pre-Auth Code Flow
4: consider using `scope` in Credential Request

### PR DESCRIPTION
**Pull request by paulbastian**
Closes #294 

- [x] make credential_identifiers mandatory in credential request for RAR (can still use format/credential_config_id in authorization req)
- [x] enable returning credential_identifiers from the token endpoint for Pre-Auth Code Flow when RAR is used
- [x] enable authorization_details with credential_configuration_id in Token Request for Pre-Auth Code Flow (already enabled by RAR)
- [ ] discuss whether to enable credential_identifiers for non-authorization_details flow as well or pit this to separate PR

There is a separate discussion to make `credential_identfiiers` available to scope flow as well, I would try to keep the discussion separate from this PR

### CODE REVIEW

File: openid-4-verifiable-credential-issuance-1_0.md
Comment (paulbastian) [inline openid-4-verifiable-credential-issuance-1_0.md:1086]:
```suggestion
  * `format`: REQUIRED if the `credential_identifiers` parameter was not returned from the Token Response. It MUST NOT be used otherwise. See (#credential-request).
```

File: openid-4-verifiable-credential-issuance-1_0.md
Comment (paulbastian) [inline openid-4-verifiable-credential-issuance-1_0.md:724]:
```suggestion
* `format`: REQUIRED if the `credential_identifiers` parameter was not returned from the Token Response. It MUST NOT be used otherwise. String that determines the format of the Credential to be issued, which may determine the type and any other information related to the Credential to be issued. Credential Format Profiles consist of the Credential format specific parameters that are defined in (#format-profiles). When this parameter is used, the `credential_identifier` Credential Request parameter MUST NOT be present.
```

File: openid-4-verifiable-credential-issuance-1_0.md
Comment (paulbastian) [reply openid-4-verifiable-credential-issuance-1_0.md:614]: I'm not sure on this one, I think the header level is already correct

File: openid-4-verifiable-credential-issuance-1_0.md
Comment (David-Chadwick) [inline openid-4-verifiable-credential-issuance-1_0.md:732]:
```suggestion
* `format`: REQUIRED if the `credential_identifiers` parameter was not returned from the Token Response. It MUST NOT be used otherwise. A string that determines the format of the Credential to be issued, which may determine the type and any other information related to the Credential to be issued. Credential Format Profiles consist of the Credential format specific parameters that are defined in (#format-profiles). When this parameter is used, the `credential_identifier` Credential Request parameter MUST NOT be present.
```

File: openid-4-verifiable-credential-issuance-1_0.md
Comment (David-Chadwick) [inline openid-4-verifiable-credential-issuance-1_0.md:733]:
```suggestion
* `credential_identifier`: REQUIRED when the `credential_identifiers` parameter was returned from the Token Response. It MUST NOT be used otherwise. A string that identifies a Credential Dataset that is requested for issuance. When this parameter is used, the `format` parameter and any other Credential format specific parameters such as those defined in (#format-profiles) MUST NOT be present.
```

File: openid-4-verifiable-credential-issuance-1_0.md
Comment (Sakurann) [reply openid-4-verifiable-credential-issuance-1_0.md:616]:
```suggestion
Credential Issuers MAY support requesting authorization to issue a Credential using the `authorization_details` parameter.
```

File: openid-4-verifiable-credential-issuance-1_0.md
Comment (Sakurann) [inline openid-4-verifiable-credential-issuance-1_0.md:618]:
```suggestion
This is particularly useful, if the Credential Issuer offered multiple Credential Configurations in the Credential Offer of a Pre-Authorized Code Flow. The Wallet can use `authorization_details` in the Token Request to request a specific Credential Configuration. The value of the `authorization_details` parameter is defined in (#(authorization-details)).
```

File: openid-4-verifiable-credential-issuance-1_0.md
Comment (Sakurann) [reply openid-4-verifiable-credential-issuance-1_0.md:644]: i would like to keep the text as-is

File: openid-4-verifiable-credential-issuance-1_0.md
Comment (Sakurann) [reply openid-4-verifiable-credential-issuance-1_0.md:647]: it is a note, so not sure a normative statement goes well with it. also, credential_identifiers is a parameter that is part of authorization details, so it should be clear that it cannot be used with scopes.

Comment (Sakurann) [reply openid-4-verifiable-credential-issuance-1_0.md:647]:
```suggestion
Note: `credential_identifiers` parameters cannot be used when the `scope` parameter is used in the Authorization Request to request issuance of a Credential.
```

File: openid-4-verifiable-credential-issuance-1_0.md
Comment (Sakurann) [inline openid-4-verifiable-credential-issuance-1_0.md:732]: can we please stick to `REQUIRED when` or `REQUIRED if` throughout the specification text?

File: openid-4-verifiable-credential-issuance-1_0.md
Comment (Sakurann) [inline openid-4-verifiable-credential-issuance-1_0.md:2482]:
```suggestion
   * clarify that authorization_details can be present in the Token Request for Pre-Authorized Code Flow when multiple Credential Configurations are present in the Credential Offer
```
I cannot interpret the current sentence..

File: openid-4-verifiable-credential-issuance-1_0.md
Comment (Sakurann) [inline openid-4-verifiable-credential-issuance-1_0.md:618]:
```suggestion
Credential Issuers MAY support requesting authorization to issue a Credential using the authorization_details parameter. This is particularly useful, if the Credential Issuer offered multiple Credential Configurations in the Credential Offer of a Pre-Authorized Code Flow. 

The Wallet may use `authorization_details` in the Token Request to request a specific Credential Configuration. The value of the `authorization_details` parameter is defined in (#(authorization-details)).
```
I think this makes the point of this sentence clearer.

**Review by Sakurann — COMMENTED** [review]
I would like to see this point made clearer (made suggestions):

> 3: enable authorization_details with credential_configuration_id in Token Request for Pre-Auth Code Flow (already enabled by RAR)

File: openid-4-verifiable-credential-issuance-1_0.md
Comment (peppelinux) [inline openid-4-verifiable-credential-issuance-1_0.md:616]:
```suggestion
Credential Issuers MAY support requesting authorization to issue a Credential using the `authorization_details` parameter. This is particularly useful, if the Credential Issuer offered multiple Credential Configurations in the Credential Offer of a Pre-Authorized Code Flow. 
```

File: openid-4-verifiable-credential-issuance-1_0.md
Comment (peppelinux) [inline openid-4-verifiable-credential-issuance-1_0.md:2487]:
```suggestion
   * clarify optionality of scope and `authorization_details` for Authorization Request
```

**Review by jogu — APPROVED** [review]

File: openid-4-verifiable-credential-issuance-1_0.md
Comment (jogu) [inline openid-4-verifiable-credential-issuance-1_0.md:733]: It's not a problem introduced in this pull request, but I don't find the "the `credential_identifiers` parameter was returned from the Token Response" language clear, and in fact on first reading it confused me into thinking we were talking about a new top level element in the token endpoint response.

Language that would be consistent with https://www.rfc-editor.org/rfc/rfc9396.html would be something like:

```suggestion
* `credential_identifier`: REQUIRED when an Authorization Details of type `openid_credential` containing the `credential_identifiers` field was returned in the Token Response. It MUST NOT be used otherwise. A string that identifies a Credential Dataset that is requested for issuance. When this parameter is used, the `format` parameter and any other Credential format specific parameters such as those defined in (#format-profiles) MUST NOT be present.
```

(If we agree this is better language then there's other places it should be applied too.)

File: openid-4-verifiable-credential-issuance-1_0.md
Comment (David-Chadwick) [inline openid-4-verifiable-credential-issuance-1_0.md:618]:
```suggestion
The Wallet can use `authorization_details` in the Token Request to request a specific Credential Configuration in both the Authorization Code Flow and the Pre-Authorized Code Flow. The value of the `authorization_details` parameter is defined in (#(authorization-details)).
```

File: openid-4-verifiable-credential-issuance-1_0.md
Comment (David-Chadwick) [inline openid-4-verifiable-credential-issuance-1_0.md:644]:
```suggestion
* `authorization_details`: REQUIRED when the `authorization_details` parameter is used to request issuance of a certain Credential Configuration as defined in (#authorization-details). It MUST NOT be used otherwise. It is an array of objects, as defined in Section 7 of [@!RFC9396]. In addition to the parameters defined in (#authorization-details), this specification defines the following parameter to be used with the authorization details type `openid_credential` in the Token Response:
```

File: openid-4-verifiable-credential-issuance-1_0.md
Comment (David-Chadwick) [inline openid-4-verifiable-credential-issuance-1_0.md:647]:
```suggestion
Note: `credential_identifiers` cannot be used when the `scope` parameter is used in the Authorization Request to request issuance of a Credential.
```

Comment (David-Chadwick) [reply openid-4-verifiable-credential-issuance-1_0.md:647]: either the above or " the 'credential_identifiers' parameter.....

**Review by David-Chadwick — APPROVED** [review]
Subject to minor editorials

**Comment by TakahikoKawasaki** [comment]
How does a user approve the value of the `authorization_details` request parameter in a token request?

For example, when a client application makes an authorization request without the `authorization_details` request parameter but includes the `authorization_details` request parameter with `{"type":"openid_credential",...}` in the subsequent token request, should the authorization server issue an access token that can be used for verifiable credential issuance? If the authorization server issues such an access token in this case, client applications can obtain an access token with arbitrary `authorization_details` without user approval.

**Comment by jogu** [comment]
> How does a user approve the value of the `authorization_details` request parameter in a token request?
> 
> For example, when a client application makes an authorization request without the `authorization_details` request parameter but includes the `authorization_details` request parameter with `{"type":"openid_credential",...}` in the subsequent token request, should the authorization server issue an access token that can be used for verifiable credential issuance? If the authorization server issues such an access token in this case, client applications can obtain an access token with arbitrary `authorization_details` without user approval.

My understanding is the authorization server should return an access token with no more than the permissions that were already granted. The mechanism shouldn't be used to obtain new permissions without user approval.

**Comment by paulbastian** [comment]
> > How does a user approve the value of the `authorization_details` request parameter in a token request?
> > For example, when a client application makes an authorization request without the `authorization_details` request parameter but includes the `authorization_details` request parameter with `{"type":"openid_credential",...}` in the subsequent token request, should the authorization server issue an access token that can be used for verifiable credential issuance? If the authorization server issues such an access token in this case, client applications can obtain an access token with arbitrary `authorization_details` without user approval.
> 
> My understanding is the authorization server should return an access token with no more than the permissions that were already granted. The mechanism shouldn't be used to obtain new permissions without user approval.

That's my view as well, authorization_details in Token Request is only really applicable to PreAuthCode Flow.

File: openid-4-verifiable-credential-issuance-1_0.md
Comment (paulbastian) [reply openid-4-verifiable-credential-issuance-1_0.md:733]: I don't like the current language as well, especially for the `format`: "REQUIRED if the `credential_identifiers` parameter was not returned from the Token Response", which seems very ugly

Comment (jogu) [reply openid-4-verifiable-credential-issuance-1_0.md:733]: Discussed on today's WG call; as `credential_identifiers` is required there's no need to mention it:

```suggestion
* `credential_identifier`: REQUIRED when an Authorization Details of type `openid_credential` was returned from the Token Response. It MUST NOT be used otherwise. A string that identifies a Credential Dataset that is requested for issuance. When this parameter is used, the `format` parameter and any other Credential format specific parameters such as those defined in (#format-profiles) MUST NOT be present.
```

File: openid-4-verifiable-credential-issuance-1_0.md
Comment (jogu) [inline openid-4-verifiable-credential-issuance-1_0.md:732]: Discussed on today's WG call how to improve this language, this was the consensus:

```suggestion
* `format`: REQUIRED if an Authorization Details of type `openid_credential` was not returned from the Token Response (e.g. when the credential was requested using a `scope` value in the authorization request or a pre-authorisation code was used that did not return an Authorization Details). It MUST NOT be used otherwise. A string that determines the format of the Credential to be issued, which may determine the type and any other information related to the Credential to be issued. Credential Format Profiles consist of the Credential format specific parameters that are defined in (#format-profiles). When this parameter is used, the `credential_identifier` Credential Request parameter MUST NOT be present.
```

**Comment by jogu** [comment]
We discussed this PR on today's WG call and seemed to have consensus that this was good to merge once the two suggestions I posted above are applied (and the git conflicts fixed!).

**Comment by paulbastian** [comment]
PR was discussed at latest DCP WG Call, I made the proposed changes and PR is ready from my side

**Comment by jogu** [comment]
3 approvals, open for over a week, all comments addressed, discussed on last week's wg call - merging!

**Pull request merged**
